File: lua/binary-search.lua
return function(array, target)
    local i = 1
    local j = #array
    while i <= j do
        local m = (i + j) // 2
        if array[m] == target then
            return m
        elseif array[m] < target then
            i = m + 1
        else
            j = m - 1
        end
    end
    return -1
end

File: zig/binary_search.zig
// Take a look at the tests, you might have to change the function arguments

pub fn binarySearch(comptime T: type, target: T, items: []const T) ?usize {
    if (items.len == 0) {
        return null;
    }
    var i: usize = 0;
    var j: usize = items.len - 1;
    while (i <= j) {
        const m = i + (j - i) / 2;
        if (items[m] == target) {
            return m;
        }
        if (items[m] < target) {
            i = m + 1;
        } else if (m > 0) {
            j = m - 1;
        } else {
            return null;
        }
    }
    return null;
}
